src/control/server/server.go

```diff
@@ -163,16 +164,23 @@ func (srv *server) createServices(ctx context.Context) error {
 // OnEnginesStarted adds callback functions to be called when all engines have
 // started up.
 func (srv *server) OnEnginesStarted(fns ...func(context.Context) error) {
+	srv.cbLock.Lock()
 	srv.onEnginesStarted = append(srv.onEnginesStarted, fns...)
+	srv.cbLock.Unlock()
 }
 
 // OnShutdown adds callback functions to be called when the server shuts down.
 func (srv *server) OnShutdown(fns ...func()) {
+	srv.cbLock.Lock()
 	srv.onShutdown = append(srv.onShutdown, fns...)
+	srv.cbLock.Unlock()
 }
 
 func (srv *server) shutdown() {
-	for _, fn := range srv.onShutdown {
+	srv.cbLock.Lock()
+	onShutdownCbs := srv.onShutdown
+	srv.cbLock.Unlock()
```

Comment:
Heh, I'd completely forgotten that I added this mutex to quiet the race detector warning on shutdown.

So, yes. Assigning a slice variable to another variable doesn't make a copy of the slice, it's just copying the pointer to the slice. I think that this is safe in this case because what we're guarding against is unsafe concurrent access to the `srv.onShutdown` field. When `srv.OnShutdown()` is called, a new slice is created using `append()` and the pointer to that slice is saved in `srv.onShutdown`.

The reason that I'm doing the assignment under the lock rather than the loop is that I was concerned that a callback could do something that also needs the lock and then we've got a deadlock situation.

Reply:
Ah, got it. Thanks. This should resolve DAOS-7809 as well.